_posts/2021-01-25-troubleshooting-puppetized-dsc-modules.md

```diff
@@ -0,0 +1,348 @@
+
+### CIM Instances
+
+Some DSC Resources use nested CIM instances to pass hashes as property values in these cases, there's one or more lines above the definition of the `InvokeParams` variable which creates these CIM instances;
+this is necessary for DSC to correctly handle the objects, _especially_ if they have a CIM instance class specific to the DSC Resource being called.
+
+```powershell
+$b615d113_bb6e_49e4_9776_7e895dfe20c7 = New-CimInstance -ClientOnly -ClassName 'MSFT_KeyValuePair' -Property @{'key' = 'Accept-Language' ; 'value' = 'en-US'}
+$InvokeParams = @{
+  Name = 'xRemoteFile';
+  Method = 'set';
+  Property = @{
+    destinationpath = 'C:\dsc-xpsdesiredstateconfiguration-9.1.0-0-1.tar.gz';
+    headers = [CimInstance[]]@($b615d113_bb6e_49e4_9776_7e895dfe20c7);
+    uri = 'https://forge.puppet.com/v3/files/dsc-xpsdesiredstateconfiguration-9.1.0-0-1.tar.gz'
+  };
+  ModuleName = @{
+    ModuleName = 'C:/pathToPuppetModules/xpsdesiredstateconfiguration/lib/puppet_x/xpsdc/dsc_resources/xPSDesiredStateConfiguration/xPSDesiredStateConfiguration.psd1';
+    RequiredVersion = '9.1.0'
+  }
+}
+```
+
+## Digging deeper
+
+Sometimes, reading the debug logs doesn't give you enough control over what's going on.
+When this is the case, the most powerful thing you can do is learn to interactively debug a Puppet run.
+
+### Getting started with Pry
+
+[Pry][pry] is a powerful tool for interactively debugging ruby.
+Using it with Puppet on Windows is a bit involved but it can be an extremely useful tool in your kit.
+
+First, you'll want to get a functional ruby environment installed _separate_ from the PDK - right now, the PDK does not support prying into a Puppet run due to limitations around native gems.
+
+We suggest you do this via [Glenn Sarti's `RubyInstaller`][rubyinstaller] PowerShell module:
+
+```powershell
+Install-Module RubyInstaller -Scope CurrentUser
+# This will kick off an interactive installation of Ruby, including the necessary dependencies
+# It relies on Chocolatey (and will install it if needed)
+# Strongly suggest installation option G ('2.5.1-2 (x64)') at the time of this writing.
+Install-Ruby
+```
+
+The next step is to find the module files on disk.
+They're _probably_ in a location like this one:
+
+```
+C:/ProgramData/PuppetLabs/code/environments/<your environment name>/modules/
+```
+
+You're going to want to `Set-Location` to the module you want to debug, like:
+
+```powershell
+Set-Location 'C:/ProgramData/PuppetLabs/code/environments/production/modules/powershellget'
+```
+
+We strongly suggest opening this location in VSCode to make editing a bit easier on yourself.
+In any case, you will want to _delete_ the file `Gemfile.lock` in this directory and create a new file, `Gemfile.local`, with the following content:
+
+```ruby
+gem 'fuubar'
+gem 'pry-byebug'
+gem 'pry-stack_explorer'
+```
+
+Next, you're going to need to run the following commands:
+
+```
```

Comment:
The way this renders in Jekyll is a bit hard to read. Would you be open to the suggestion of splitting this up in to four individual blocks:
```
# This loads ruby 2.5 for use; if you don't do this, you won't be able to run other commands
uru 2.5
```
```
# This is only necessary after you first install a ruby version; it's a tool to manage ruby dependencies
gem install bundler
```

etc...

Reply:
Sure, will rework for this.

Reply:
Hopefully this works?